src/coordinator/coordinator_server.h

```diff
@@ -0,0 +1,100 @@
+};
+
+class CoordinatorServer : public Coordinator::Service {
+ public:
+  explicit CoordinatorServer(absl::Duration default_presumed_abort_duration)
+      : default_presumed_abort_duration_(default_presumed_abort_duration) {}
+
+  grpc::Status CommitAtomicTransaction(
+      grpc::ServerContext *context,
+      const CommitAtomicTransactionRequest *request,
+      CommitAtomicTransactionResponse *response) override;
+
+  grpc::Status GetTransactionResult(
+      grpc::ServerContext *context, const GetTransactionResultRequest *request,
+      GetTransactionResultResponse *response) override;
+
+  // These are protected so they can be mocked out in testing.
+ protected:
+  virtual void PrepareCohortTransaction(
+      const std::string &transaction_id, const common::Namespace &ns,
+      const cohort::PrepareTransactionRequest &request,
+      const grpc::ServerContext &context);
+  virtual std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+      cohort::GetTransactionResultResponse>>
+  AsyncGetResultsFromCohort(const common::Namespace &ns,
+                            const cohort::GetTransactionResultRequest &request,
+                            grpc::ClientContext &context);
+  virtual grpc::Status FinishAsyncGetResultsFromCohort(
+      std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+          cohort::GetTransactionResultResponse>> &async_response,
+      cohort::GetTransactionResultResponse &response);
+  virtual absl::Time Now();
+  virtual absl::Status StartVoting(
+      const std::string &transaction_id,
+      const google::protobuf::Timestamp &presumed_abort_time,
+      size_t num_cohorts);
+  virtual absl::StatusOr<blockchain::TwoPhaseCommit::VotingDecision>
+  GetVotingDecision(const std::string &transaction_id);
+
+ private:
+  virtual cohort::Cohort::StubInterface &GetCohortStub(
+      const common::Namespace &ns);
+  grpc::Status UpdateResponseForSingleCohortTransaction(
+      const std::string &transaction_id, const grpc::ServerContext &context);
+  // Updates response to client when the blockchain says the transaction
+  // aborted.
+  void UpdateAbortedResponseFromCohorts(const std::string &transaction_id,
+                                        const grpc::ServerContext &context);
+  // Updates response to client when the blockchain says the transaction
+  // committed.
+  void UpdateCommittedResponseFromCohorts(const std::string &transaction_id,
+                                          const grpc::ServerContext &context);
+  // Garbage collects metadata for a transaction once the final response is
+  // known.
+  void CleanUpTransactionMetadata(const std::string &transaction_id);
+
+  absl::flat_hash_map<std::string,
+                      std::unique_ptr<cohort::Cohort::StubInterface>>
+      cohort_by_namespace_;
+  absl::flat_hash_map<std::string, TransactionMetadata>
+      metadata_by_transaction_;
+  absl::flat_hash_map<std::string, TransactionResponse>
+      response_by_transaction_;
```

Comment:
Is it possible that any of these states are accessed by different thread (e.g. client and cohorts)? Do we need add locks to protect them?

Reply:
Yeah I should add locks. That's a good catch. We should ensure that only one thread can currently be accessing any metadata of a single transaction at any given time.